File: 프로그래머스/lv2/87946. 피로도/피로도.py
answer = 0
visited = []

def solution(k, dungeons):
    dun_num = len(dungeons)
    visited = [False] * len(dungeons)
    
    def dfs(k, dungeons, depth):
        global answer
        answer = max(answer, depth)

        for i in range(len(dungeons)):
            if not visited[i] and dungeons[i][0] <= k:
                visited[i] = True
                dfs(k - dungeons[i][1], dungeons, depth + 1)
                visited[i] = False
    dfs(k, dungeons, 0)
    
    return answer

File: 프로그래머스/lv2/87946. 피로도/피로도.java
class Solution {
    private static int answer;
    private static boolean[] visited;
    
    public int solution(int k, int[][] dungeons) {
        answer = 0;
        visited = new boolean[dungeons.length];
        dfs(dungeons, 0, k);

        return answer;
    }

    public static void dfs(int[][] dungeons, int depth, int k) {
        if (depth > dungeons.length) {
            return;
        }

        for (int i = 0; i < dungeons.length; i++) {
            if (!visited[i] && k >= dungeons[i][0]) {
                visited[i] = true;
                dfs(dungeons, depth + 1, k - dungeons[i][1]);
                visited[i] = false;
            }
        }
        answer = Math.max(answer, depth);
    }
}